packages/contracts/contracts/optimistic-ethereum/OVM/bridge/messaging/OVM_L1CrossDomainMessenger.sol

```diff
@@ -70,14 +101,58 @@ contract OVM_L1CrossDomainMessenger is iOVM_L1CrossDomainMessenger, Abs_BaseCros
         address _libAddressManager
     )
         public
+        initializer
     {
         require(
             address(libAddressManager) == address(0),
             "L1CrossDomainMessenger already intialized."
         );
-
         libAddressManager = Lib_AddressManager(_libAddressManager);
         xDomainMsgSender = DEFAULT_XDOMAIN_SENDER;
+
+        // Initialize upgradable OZ contracts
+        __Context_init_unchained(); // Context is a dependency for both Ownable and Pausable
+        __Ownable_init_unchained();
+        __Pausable_init_unchained();
+        __ReentrancyGuard_init_unchained();
+    }
+
+    /**
+     * Pause relaying.
+     */
+    function pause()
+        external
+        onlyOwner
+    {
+        _pause();
+    }
+
+    /**
+     * Block a message.
+     * @param _xDomainCalldataHash Hash of the message to block.
+     */
+    function blockMessage(
+        bytes32 _xDomainCalldataHash
+    )
+        external
+        onlyOwner
+    {
+        blockedMessages[_xDomainCalldataHash] = true;
+        emit MessageBlocked(_xDomainCalldataHash);
+    }
+
+    /**
+     * Allow a message.
+     * @param _xDomainCalldataHash Hash of the message to block.
+     */
+    function allowMessage(
+        bytes32 _xDomainCalldataHash
+    )
+        external
+        onlyOwner
+    {
+        blockedMessages[_xDomainCalldataHash] = false;
```

Comment:
Have you thought about not adding a `blockedMessages` mapping and instead just writing to `successfulMessages`? This would save a `SLOAD` in `relayMessage`. The name `successfulMessages` doesn't really make sense anymore if that was the case, it could be something like `messages` or `relayedMessages`

Reply:
Good question! Yes, it's mentioned in the PR description, and called out [inline here](https://github.com/ethereum-optimism/optimism/pull/579#discussion_r618915230). 

An important downside IMO is that it changes the trust assumptions significantly, because if we retain the `allowMessage()` functionality, the `owner` could delete entries, making message replayable. Separate mapping prevent that.

Reply:
Agreed that this is not a permission we want to enable.